lightning/backend/local.go

```diff
@@ -0,0 +1,1159 @@
+	ch chan []byte
+}
+
+// recycleChan is used for reusing allocated []byte so we can use memory more efficiently
+//
+// NOTE: we don't used a `sync.Pool` because when will sync.Pool release is depending on the
+// garbage collector which always release the memory so late. Use a fixed size chan to reuse
+// can decrease the memory use to 1/3 compare with sync.Pool.
+var recycleChan *bytesRecycleChan
+
+func init() {
+	recycleChan = &bytesRecycleChan{
+		ch: make(chan []byte, 1024),
+	}
+}
+
+func (c *bytesRecycleChan) Acquire() []byte {
+	select {
+	case b := <-c.ch:
+		return b
+	default:
+		return manual.New(2 << 20) // 1M
+	}
+}
+
+func (c *bytesRecycleChan) Release(w []byte) {
+	select {
+	case c.ch <- w:
+		return
+	default:
+		manual.Free(w)
+	}
+}
+
+type bytesBuffer struct {
+	bufs      [][]byte
+	curBuf    []byte
+	curIdx    int
+	curBufIdx int
+	curBufLen int
+}
+
+func newBytesBuffer() *bytesBuffer {
+	return &bytesBuffer{bufs: make([][]byte, 0, 128), curBufIdx: -1}
+}
+
+func (b *bytesBuffer) addBuf() {
+	if b.curBufIdx < len(b.bufs)-1 {
+		b.curBufIdx += 1
+		b.curBuf = b.bufs[b.curBufIdx]
+	} else {
+		buf := recycleChan.Acquire()
+		b.bufs = append(b.bufs, buf)
+		b.curBuf = buf
+		b.curBufIdx = len(b.bufs) - 1
+	}
+
+	b.curBufLen = len(b.curBuf)
+	b.curIdx = 0
+}
+
+func (b *bytesBuffer) reset() {
+	if len(b.bufs) > 0 {
+		b.curBuf = b.bufs[0]
+		b.curBufLen = len(b.bufs[0])
+		b.curBufIdx = 0
+		b.curIdx = 0
+	}
+}
+
+func (b *bytesBuffer) destroy() {
+	for _, buf := range b.bufs {
+		recycleChan.Release(buf)
+	}
+	b.bufs = b.bufs[:0]
+}
+
+func (b *bytesBuffer) totalSize() int64 {
+	return int64(len(b.bufs)) * int64(1<<20)
+}
+
+func (b *bytesBuffer) addBytes(bytes []byte) []byte {
+	if len(bytes) > bigValueSize {
+		return append([]byte{}, bytes...)
+	} else {
+		if b.curIdx+len(bytes) > b.curBufLen {
+			b.addBuf()
+		}
+		idx := b.curIdx
+		copy(b.curBuf[idx:], bytes)
+		b.curIdx += len(bytes)
+		return b.curBuf[idx:b.curIdx]
+	}
+}
+
+func (local *local) writeAndIngestByRange(
+	ctx context.Context,
+	engineFile *LocalFile,
+	start, end []byte) error {
+	select {
+	case <-ctx.Done():
+		return errors.New("context is cancel by other reason")
+	default:
+	}
+
+	ito := &pebble.IterOptions{
+		LowerBound: start,
+		UpperBound: end,
+	}
+
+	iter := engineFile.db.NewIter(ito)
+	defer iter.Close()
+	// Needs seek to first because NewIter returns an iterator that is unpositioned
+	hasKey := iter.First()
+	if !hasKey {
+		log.L().Info("There is no pairs in iterator",
+			zap.Binary("start", start),
+			zap.Binary("end", end),
+			zap.Binary("next end", nextKey(end)))
+		return nil
+	}
+	pairStart := append([]byte{}, iter.Key()...)
+	iter.Last()
+	pairEnd := append([]byte{}, iter.Key()...)
+
+	var regions []*split.RegionInfo
+	var err error
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+WriteAndIngest:
+	for retry := 0; retry < maxRetryTimes; retry++ {
+		if retry != 0 {
+			time.Sleep(time.Second)
```

Comment:
This sleep should be `select`ed

```suggestion
			select {
			case <-time.After(time.Second):
			case <-ctx.Done():
				return ctx.Err()
			}
```